WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m

```diff
@@ -0,0 +1,102 @@
+ */
+
+#import "XCUIElement+FBClassChain.h"
+#import "FBClassChainQueryParser.h"
+#import "FBElementUtils.h"
+#import "XCElementSnapshot.h"
+#import "XCUIElement+FBUtilities.h"
+
+NSString *const FBClassChainQueryParseException = @"FBClassChainQueryParseException";
+
+@implementation XCUIElement (FBClassChain)
+
+- (NSArray<XCUIElement *> *)fb_descendantsMatchingClassChain:(NSString *)classChainQuery shouldReturnAfterFirstMatch:(BOOL)shouldReturnAfterFirstMatch
+{
+  NSError *error;
+  FBClassChain parsedChain = [FBClassChainQueryParser parseQuery:classChainQuery error:&error];
+  if (nil == parsedChain) {
+    @throw [NSException exceptionWithName:FBClassChainQueryParseException reason:error.localizedDescription userInfo:error.userInfo];
+    return nil;
+  }
+  NSMutableArray *unmatchedSnapshots = [NSMutableArray arrayWithArray:[self fb_lookupChain:parsedChain root:self.fb_lastSnapshot]];
+  if (0 == unmatchedSnapshots.count) {
+    return @[];
+  }
+  if (shouldReturnAfterFirstMatch) {
+    [unmatchedSnapshots removeObjectsInRange:NSMakeRange(1, unmatchedSnapshots.count - 1)];
+  }
+  XCUIElementQuery *elementsQuery = [self childrenMatchingType:[parsedChain firstObject].type];
+  for (NSUInteger level = 1; level < parsedChain.count; ++level) {
+    elementsQuery = [elementsQuery childrenMatchingType:[parsedChain objectAtIndex:level].type];
+  }
+  NSMutableArray *result = [NSMutableArray array];
+  NSMutableArray *candidateElements = [NSMutableArray arrayWithArray:elementsQuery.allElementsBoundByIndex];
+  if ([parsedChain lastObject].position < 0 && candidateElements.count > 1) {
+    candidateElements = [[[candidateElements reverseObjectEnumerator] allObjects] mutableCopy];
+  }
+  NSUInteger candidateElementIdx = 0;
+  while (candidateElementIdx < candidateElements.count) {
+    XCUIElement *candidateElement = [candidateElements objectAtIndex:candidateElementIdx];
+    NSUInteger snapshotIdx = 0;
+    BOOL isMatchFound = NO;
+    while (snapshotIdx < unmatchedSnapshots.count) {
+      if ([candidateElement.fb_lastSnapshot _matchesElement:[unmatchedSnapshots objectAtIndex:snapshotIdx]]) {
+        [result addObject:candidateElement];
+        [unmatchedSnapshots removeObjectAtIndex:snapshotIdx];
+        break;
+      }
+      ++snapshotIdx;
+    }
+    if (0 == unmatchedSnapshots.count) {
+      break;
+    }
+    if (isMatchFound) {
+      [candidateElements removeObjectAtIndex:candidateElementIdx];
+    } else {
+      ++candidateElementIdx;
+    }
+  }
+  return result.copy;
+}
+
+- (NSArray<XCElementSnapshot *> *)fb_lookupChain:(FBClassChain)query root:(XCElementSnapshot *)root
+{
+  FBClassChainElement *chainElement = [query firstObject];
+  NSArray *filteredChildren = root.children;
+  if (XCUIElementTypeAny != chainElement.type) {
+    filteredChildren = [filteredChildren filteredArrayUsingPredicate:[NSPredicate predicateWithFormat:@"elementType == %@", @(chainElement.type)]];
```

Comment:
Please use `FBStringify` instead of plain strings, https://github.com/facebook/WebDriverAgent/blob/2eefc210ab3a775cfd1eff1b47e1a81484b64b50/WebDriverAgentLib/Commands/FBFindElementCommands.m#L77